[Bug]: App crashes when adding voice recording due to missing speech recognition activity

**Bug Description**
The app crashes when trying to add a voice recording to a note. This happens due to an unhandled `ActivityNotFoundException` when attempting to launch the speech recognition activity.

**To Reproduce**
Steps to reproduce the behavior:

1. Tap on "Create New Note" button.
2. Add a voice recording for the new note.
3. The app crashes immediately.

**Expected Behavior**
The app should either:
- Successfully launch a voice recording interface, or
- Gracefully handle the absence of a speech recognition activity (e.g., show an error message or disable the feature if unsupported).

**Screen Recording**
https://github.com/user-attachments/assets/e3554d99-f46d-44f0-9cff-7944fa1d51c8

**Smartphone:**

- Device: generic_x86_64 (Android Studio x86 emulator)
- OS: Android 9.0 (Default Android System Image)
- Version: 1.3.7

**Suggested Fix**
Check if there's an app available to handle the `android.speech.action.RECOGNIZE_SPEECH` intent before launching it. If not, inform the user or disable the feature to avoid a crash.

**Error Log**
```
FATAL EXCEPTION: main
Process: com.mshdabiola.playnotepad.foss, PID: 3127
java.lang.RuntimeException: Failure delivering result ResultInfo{...} to activity {...MainActivity}: 
android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.speech.action.RECOGNIZE_SPEECH (has extras) }
...
Caused by: android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.speech.action.RECOGNIZE_SPEECH (has extras) }
	at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:2007)
	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1673)
	...
```

--- a/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
+++ b/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
@@ -58,6 +58,7 @@ import com.mshdabiola.playnotepad.navigation.NoteNavHost
 import com.mshdabiola.setting.navigation.navigateToSetting
 import com.mshdabiola.ui.AudioDialog
 import com.mshdabiola.ui.ImageDialog2
+import com.mshdabiola.ui.supportVoice
 import kotlinx.coroutines.launch
 
 @OptIn(ExperimentalComposeUiApi::class)
@@ -165,6 +166,7 @@ fun NoteApp(
                                         appState.navController.navigateToDrawing(id.first, id.second)
                                     }
                                 },
+                                isVoiceSupport = supportVoice(),
 
                             )
                         }
@@ -246,8 +248,10 @@ fun NoteBottomBar(
     onAddDrawNote: () -> Unit = {},
     onAddVoiceNote: () -> Unit = {},
     onAddImageNote: () -> Unit = {},
+    isVoiceSupport: Boolean = false,
 ) {
     BottomAppBar(
+        modifier = modifier,
         actions = {
             IconButton(
                 modifier = Modifier.testTag("main:check"),
@@ -269,14 +273,16 @@ fun NoteBottomBar(
                 )
             }
 
-            IconButton(
-                modifier = Modifier.testTag("main:voice"),
-                onClick = onAddVoiceNote,
-            ) {
-                Icon(
-                    imageVector = NoteIcon.KeyboardVoice,
-                    contentDescription = "add note voice",
-                )
+            if (isVoiceSupport) {
+                IconButton(
+                    modifier = Modifier.testTag("main:voice"),
+                    onClick = onAddVoiceNote,
+                ) {
+                    Icon(
+                        imageVector = NoteIcon.KeyboardVoice,
+                        contentDescription = "add note voice",
+                    )
+                }
             }
 
             IconButton(

--- a/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
+++ b/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
@@ -274,16 +274,28 @@ fun NoteBottomBar(
             }
 
             if (isVoiceSupport) {
-                IconButton(
-                    modifier = Modifier.testTag("main:voice"),
-                    onClick = onAddVoiceNote,
-                ) {
-                    Icon(
-                        imageVector = NoteIcon.KeyboardVoice,
-                        contentDescription = "add note voice",
-                    )
-                }
-            }
+                        IconButton(
+                            modifier = Modifier.testTag("main:voice"),
+                            onClick = onAddVoiceNote,
+                        ) {
+                            Icon(
+                                imageVector = NoteIcon.KeyboardVoice,
+                                contentDescription = "add note voice",
+                            )
+                        }
+                    } else {
+                        IconButton(
+                            modifier = Modifier.testTag("main:voice").then(Modifier.disabled()),
+                            onClick = {}, // or show a tooltip
+                            enabled = false
+                        ) {
+                            Icon(
+                                imageVector = NoteIcon.KeyboardVoice,
+                                contentDescription = "add note voice (unavailable)",
+                                tint = Color.Gray // or other visual cue
+                            )
+                        }
+                    }
 
             IconButton(
                 modifier = Modifier.testTag("main:image"),

--- a/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
+++ b/app/src/main/java/com/mshdabiola/playnotepad/ui/NoteApp.kt
@@ -163,7 +163,10 @@ fun NoteApp(
                                     appState.coroutineScope.launch {
                                         val id = viewModel.insertNewDrawing()
                                         appState.navController.navigateToDetail(DetailArg(id.first))
-                                        appState.navController.navigateToDrawing(id.first, id.second)
+                                        appState.navController.navigateToDrawing(
+                                            id.first,
+                                            id.second,
+                                        )
                                     }
                                 },
                                 isVoiceSupport = supportVoice(),
@@ -274,28 +277,28 @@ fun NoteBottomBar(
             }
 
             if (isVoiceSupport) {
-                        IconButton(
-                            modifier = Modifier.testTag("main:voice"),
-                            onClick = onAddVoiceNote,
-                        ) {
-                            Icon(
-                                imageVector = NoteIcon.KeyboardVoice,
-                                contentDescription = "add note voice",
-                            )
-                        }
-                    } else {
-                        IconButton(
-                            modifier = Modifier.testTag("main:voice").then(Modifier.disabled()),
-                            onClick = {}, // or show a tooltip
-                            enabled = false
-                        ) {
-                            Icon(
-                                imageVector = NoteIcon.KeyboardVoice,
-                                contentDescription = "add note voice (unavailable)",
-                                tint = Color.Gray // or other visual cue
-                            )
-                        }
-                    }
+                IconButton(
+                    modifier = Modifier.testTag("main:voice"),
+                    onClick = onAddVoiceNote,
+                ) {
+                    Icon(
+                        imageVector = NoteIcon.KeyboardVoice,
+                        contentDescription = "add note voice",
+                    )
+                }
+            } else {
+                IconButton(
+                    modifier = Modifier.testTag("main:voice"),
+                    onClick = {}, // or show a tooltip
+                    enabled = false,
+                ) {
+                    Icon(
+                        imageVector = NoteIcon.KeyboardVoice,
+                        contentDescription = "add note voice (unavailable)",
+                        tint = Color.Gray, // or other visual cue
+                    )
+                }
+            }
 
             IconButton(
                 modifier = Modifier.testTag("main:image"),

--- a/feature/detail/src/main/kotlin/com/mshdabiola/detail/AddDetailBottomSheet2.kt
+++ b/feature/detail/src/main/kotlin/com/mshdabiola/detail/AddDetailBottomSheet2.kt
@@ -36,6 +36,7 @@ fun AddBottomSheet2(
     onDrawing: () -> Unit = {},
     onDismiss: () -> Unit = {},
     show: Boolean,
+    isVoiceSupport: Boolean = false,
 ) {
     val background = if (currentImage != -1) {
         NoteIcon.background[currentImage].fgColor
@@ -161,41 +162,43 @@ fun AddBottomSheet2(
                 modifier = androidx.compose.ui.Modifier.testTag("detail:drawing"),
 
             )
-            NavigationDrawerItem(
-                icon = {
-                    Icon(
-                        imageVector = NoteIcon.KeyboardVoice,
-                        contentDescription = "",
-                    )
-                },
-                label = { Text(text = stringResource(Rd.string.modules_designsystem_recording)) },
-                selected = false,
-                onClick = {
-                    onDismiss()
-                    if (context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) ==
-                        PackageManager.PERMISSION_GRANTED
-                    ) {
-                        voiceLauncher.launch(
-                            Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
-                                putExtra(
-                                    RecognizerIntent.EXTRA_LANGUAGE_MODEL,
-                                    RecognizerIntent.LANGUAGE_MODEL_FREE_FORM,
-                                )
-                                putExtra(RecognizerIntent.EXTRA_PROMPT, "Speck Now Now")
-                                putExtra(
-                                    "android.speech.extra.GET_AUDIO_FORMAT",
-                                    "audio/AMR",
-                                )
-                                putExtra("android.speech.extra.GET_AUDIO", true)
-                            },
+            if (isVoiceSupport) {
+                NavigationDrawerItem(
+                    icon = {
+                        Icon(
+                            imageVector = NoteIcon.KeyboardVoice,
+                            contentDescription = "",
                         )
-                    } else {
-                        audioPermission.launch(Manifest.permission.RECORD_AUDIO)
-                    }
-                },
-                colors = NavigationDrawerItemDefaults.colors(unselectedContainerColor = background),
-                modifier = androidx.compose.ui.Modifier.testTag("detail:recording"),
-            )
+                    },
+                    label = { Text(text = stringResource(Rd.string.modules_designsystem_recording)) },
+                    selected = false,
+                    onClick = {
+                        onDismiss()
+                        if (context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) ==
+                            PackageManager.PERMISSION_GRANTED
+                        ) {
+                            voiceLauncher.launch(
+                                Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
+                                    putExtra(
+                                        RecognizerIntent.EXTRA_LANGUAGE_MODEL,
+                                        RecognizerIntent.LANGUAGE_MODEL_FREE_FORM,
+                                    )
+                                    putExtra(RecognizerIntent.EXTRA_PROMPT, "Speck Now Now")
+                                    putExtra(
+                                        "android.speech.extra.GET_AUDIO_FORMAT",
+                                        "audio/AMR",
+                                    )
+                                    putExtra("android.speech.extra.GET_AUDIO", true)
+                                },
+                            )
+                        } else {
+                            audioPermission.launch(Manifest.permission.RECORD_AUDIO)
+                        }
+                    },
+                    colors = NavigationDrawerItemDefaults.colors(unselectedContainerColor = background),
+                    modifier = androidx.compose.ui.Modifier.testTag("detail:recording"),
+                )
+            }
             if (!isNoteCheck) {
                 NavigationDrawerItem(
                     icon = {

--- a/feature/detail/src/main/kotlin/com/mshdabiola/detail/DetailScreen.kt
+++ b/feature/detail/src/main/kotlin/com/mshdabiola/detail/DetailScreen.kt
@@ -100,6 +100,7 @@ import com.mshdabiola.ui.LabelCard
 import com.mshdabiola.ui.NotificationDialogNew
 import com.mshdabiola.ui.ReminderCard
 import com.mshdabiola.ui.TimeDialog
+import com.mshdabiola.ui.supportVoice
 import com.mshdabiola.ui.toTime
 import java.io.File
 import com.mshdabiola.designsystem.R as Rd
@@ -219,6 +220,7 @@ internal fun DetailRoute(
             navigateToDrawing(editViewModel.note.value.id, id)
         },
         onDismiss = { showModalState = false },
+        isVoiceSupport = supportVoice(),
     )
 //
     val images = note.images.map {

--- a/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
+++ b/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
@@ -1,6 +1,7 @@
 package com.mshdabiola.ui
 
 import android.Manifest
+import android.annotation.SuppressLint
 import android.content.Intent
 import android.content.pm.PackageManager
 import android.net.Uri
@@ -99,6 +100,16 @@ fun AudioDialog(
             }
         },
     )
+}
+
+@SuppressLint("QueryPermissionsNeeded")
+@Composable
+fun supportVoice(): Boolean {
+    val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
+    val context = LocalContext.current
+    val pm = context.packageManager
+    val activities = pm.queryIntentActivities(intent, 0)
+    return activities.isNotEmpty()
 }
 
 @Composable

--- a/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
+++ b/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
@@ -25,6 +25,7 @@ import androidx.compose.material3.Icon
 import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.remember
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.platform.LocalContext
@@ -104,12 +105,11 @@ fun AudioDialog(
 
 @SuppressLint("QueryPermissionsNeeded")
 @Composable
-@SuppressLint("QueryPermissionsNeeded")
-@Composable
 fun supportVoice(): Boolean {
+    val context = LocalContext.current
+
     return remember {
         val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
-        val context = LocalContext.current
         val pm = context.packageManager
         val activities = pm.queryIntentActivities(intent, 0)
         activities.isNotEmpty()

--- a/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
+++ b/modules/ui/src/main/kotlin/com/mshdabiola/ui/MainActions.kt
@@ -104,12 +104,16 @@ fun AudioDialog(
 
 @SuppressLint("QueryPermissionsNeeded")
 @Composable
+@SuppressLint("QueryPermissionsNeeded")
+@Composable
 fun supportVoice(): Boolean {
-    val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
-    val context = LocalContext.current
-    val pm = context.packageManager
-    val activities = pm.queryIntentActivities(intent, 0)
-    return activities.isNotEmpty()
+    return remember {
+        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
+        val context = LocalContext.current
+        val pm = context.packageManager
+        val activities = pm.queryIntentActivities(intent, 0)
+        activities.isNotEmpty()
+    }
 }
 
 @Composable

